rasterio/_io.pyx

```diff
@@ -746,21 +788,23 @@ cdef class DatasetReaderBase(DatasetBase):
             'boundless': boundless,
             'resampling': resampling}
 
-        # GDAL found dataset-wide alpha band or mask
-        # All band masks are equal so we can return the first
         if MaskFlags.per_dataset in self.mask_flag_enums[0]:
             return self.read_masks(1, **kwargs)
 
-        # use Alpha mask if available and looks like RGB, even if nodata is shadowing
         elif self.count == 4 and self.colorinterp[0] == ColorInterp.red:
```

Comment:
@sgillies I wonder what happens when the 4th bands is not an alpha band, will it imply that the first but can have   ColorInterp.red

Reply:
The intent of this second case is not 100% clear to me. It looks like it's for the case where we have a RGB + 4th band file where the 4th band may or may not be properly labelled "alpa" (via colorinterp) *but* the 4th band is shadowed by a nodata value. In that case, we'd fall past the first case (the first `if`) but we don't want to get to the case under `else`.

Reply:
Does that make sense @vincentsarago ?

Reply:
👌